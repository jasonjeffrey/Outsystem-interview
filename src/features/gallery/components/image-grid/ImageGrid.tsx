import React, {FunctionComponent} from 'react';
import {GridList, GridListTile} from '@material-ui/core';
import styled from 'styled-components';
import {Photo} from '../../../../api';
import {useDispatch} from 'react-redux';
import {fetchImageExifData} from '../../gallerySlice';

type ImageGridProps = {
    images?: Photo[];
    selectedIndex?: number;
};

const SelectedGridListTitle = styled(GridListTile)`
  border: 1px solid red;
`;


export const ImageGrid: FunctionComponent<ImageGridProps> = ({images = [], selectedIndex}) => {
    const dispatch = useDispatch();

    function handleImageClick(image: Photo) {
        if (image) {
            dispatch(fetchImageExifData(image));
        }
    }

    return (
        <GridList cellHeight={160} cols={3}>
            {
                images.map((image) => {
                    if (selectedIndex === image.id) {
                        return <SelectedGridListTitle key={image.id} cols={1}>
                            <img src={image.thumbnail} alt=""/>
                        </SelectedGridListTitle>
                    } else {
                        return <GridListTile
                            key={image.id}
                            data-testid={image.id}
                            cols={1}
                            onClick={() => handleImageClick(image)}>
                            <img src={image.thumbnail} alt=""/>
                        </GridListTile>
                    }
                })
            }
        </GridList>
    );
}
