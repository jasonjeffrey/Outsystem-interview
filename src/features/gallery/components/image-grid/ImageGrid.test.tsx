import {ImageGrid} from './ImageGrid';
import {render, fireEvent} from '@testing-library/react';

jest.mock('react-redux', () => ({
    useSelector: jest.fn(),
    useDispatch: jest.fn()
}));

describe('ImageGrid Component', () => {
    it('RENDERS component without any GridItems', () => {
        const component = <ImageGrid images={[]}/>
        const {asFragment} = render(component);

        expect(asFragment()).toMatchSnapshot();
    });

    it('RENDERS component with 1 GridItem', () => {
        const component = <ImageGrid images={[{id: 0, thumbnail: 'http://example.com/test.jpg'}]}/>
        const {asFragment} = render(component);

        expect(asFragment()).toMatchSnapshot();
    });

    it('RENDERS component with 1 GridItem which is selected', () => {
        const component = <ImageGrid images={[{id: 0, thumbnail: 'http://example.com/test.jpg'}]} selectedIndex={0} />
        const {asFragment} = render(component);

        expect(asFragment()).toMatchSnapshot();
    });
});
