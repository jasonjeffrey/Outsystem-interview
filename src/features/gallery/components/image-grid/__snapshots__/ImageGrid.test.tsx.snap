// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGrid Component RENDERS component with 1 GridItem 1`] = `
<DocumentFragment>
  <ul
    class="MuiGridList-root"
    style="margin: -2px;"
  >
    <li
      class="MuiGridListTile-root"
      data-testid="0"
      style="width: 33.333333333333336%; height: 164px; padding: 2px;"
    >
      <div
        class="MuiGridListTile-tile"
      >
        <img
          alt=""
          src="http://example.com/test.jpg"
        />
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ImageGrid Component RENDERS component with 1 GridItem which is selected 1`] = `
<DocumentFragment>
  .c0 {
  border: 1px solid red;
}

<ul
    class="MuiGridList-root"
    style="margin: -2px;"
  >
    <li
      class="MuiGridListTile-root c0"
      style="width: 33.333333333333336%; height: 164px; padding: 2px;"
    >
      <div
        class="MuiGridListTile-tile"
      >
        <img
          alt=""
          src="http://example.com/test.jpg"
        />
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ImageGrid Component RENDERS component without any GridItems 1`] = `
<DocumentFragment>
  <ul
    class="MuiGridList-root"
    style="margin: -2px;"
  />
</DocumentFragment>
`;
